// hexo/dist/models/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tag = exports.PostTag = exports.PostCategory = exports.PostAsset = exports.Post = exports.Page = exports.Data = exports.Category = exports.Cache = exports.Asset = void 0;
var asset_1 = require("./asset");
Object.defineProperty(exports, "Asset", { enumerable: true, get: function () { return __importDefault(asset_1).default; } });
var cache_1 = require("./cache");
Object.defineProperty(exports, "Cache", { enumerable: true, get: function () { return __importDefault(cache_1).default; } });
var category_1 = require("./category");
Object.defineProperty(exports, "Category", { enumerable: true, get: function () { return __importDefault(category_1).default; } });
var data_1 = require("./data");
Object.defineProperty(exports, "Data", { enumerable: true, get: function () { return __importDefault(data_1).default; } });
var page_1 = require("./page");
Object.defineProperty(exports, "Page", { enumerable: true, get: function () { return __importDefault(page_1).default; } });
var post_1 = require("./post");
Object.defineProperty(exports, "Post", { enumerable: true, get: function () { return __importDefault(post_1).default; } });
var post_asset_1 = require("./post_asset");
Object.defineProperty(exports, "PostAsset", { enumerable: true, get: function () { return __importDefault(post_asset_1).default; } });
var post_category_1 = require("./post_category");
Object.defineProperty(exports, "PostCategory", { enumerable: true, get: function () { return __importDefault(post_category_1).default; } });
var post_tag_1 = require("./post_tag");
Object.defineProperty(exports, "PostTag", { enumerable: true, get: function () { return __importDefault(post_tag_1).default; } });
var tag_1 = require("./tag");
Object.defineProperty(exports, "Tag", { enumerable: true, get: function () { return __importDefault(tag_1).default; } });
