// hexo/dist/extend/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tag = exports.Renderer = exports.Processor = exports.Migrator = exports.Injector = exports.Highlight = exports.Helper = exports.Generator = exports.Filter = exports.Deployer = exports.Console = void 0;
var console_1 = require("./console");
Object.defineProperty(exports, "Console", { enumerable: true, get: function () { return __importDefault(console_1).default; } });
var deployer_1 = require("./deployer");
Object.defineProperty(exports, "Deployer", { enumerable: true, get: function () { return __importDefault(deployer_1).default; } });
var filter_1 = require("./filter");
Object.defineProperty(exports, "Filter", { enumerable: true, get: function () { return __importDefault(filter_1).default; } });
var generator_1 = require("./generator");
Object.defineProperty(exports, "Generator", { enumerable: true, get: function () { return __importDefault(generator_1).default; } });
var helper_1 = require("./helper");
Object.defineProperty(exports, "Helper", { enumerable: true, get: function () { return __importDefault(helper_1).default; } });
var syntax_highlight_1 = require("./syntax_highlight");
Object.defineProperty(exports, "Highlight", { enumerable: true, get: function () { return __importDefault(syntax_highlight_1).default; } });
var injector_1 = require("./injector");
Object.defineProperty(exports, "Injector", { enumerable: true, get: function () { return __importDefault(injector_1).default; } });
var migrator_1 = require("./migrator");
Object.defineProperty(exports, "Migrator", { enumerable: true, get: function () { return __importDefault(migrator_1).default; } });
var processor_1 = require("./processor");
Object.defineProperty(exports, "Processor", { enumerable: true, get: function () { return __importDefault(processor_1).default; } });
var renderer_1 = require("./renderer");
Object.defineProperty(exports, "Renderer", { enumerable: true, get: function () { return __importDefault(renderer_1).default; } });
var tag_1 = require("./tag");
Object.defineProperty(exports, "Tag", { enumerable: true, get: function () { return __importDefault(tag_1).default; } });
